enterprise/server/hit_tracker_client/hit_tracker_client.go

```diff
@@ -2,57 +2,280 @@ package hit_tracker_client
 
 import (
 	"context"
+	"flag"
+	"sync"
 	"time"
 
-	"github.com/buildbuddy-io/buildbuddy/server/environment"
 	"github.com/buildbuddy-io/buildbuddy/server/interfaces"
+	"github.com/buildbuddy-io/buildbuddy/server/metrics"
 	"github.com/buildbuddy-io/buildbuddy/server/real_environment"
+	"github.com/buildbuddy-io/buildbuddy/server/util/authutil"
+	"github.com/buildbuddy-io/buildbuddy/server/util/grpc_client"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/durationpb"
 
+	hitpb "github.com/buildbuddy-io/buildbuddy/proto/hit_tracker"
 	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
+	rspb "github.com/buildbuddy-io/buildbuddy/proto/resource"
+	gstatus "google.golang.org/grpc/status"
 )
 
+const (
+	initialSliceCapacity = 4_096
+)
+
+var (
+	remoteHitTrackerTarget       = flag.String("cache.remote_hit_tracker.target", "", "The gRPC target of the remote cache-hit-tracking service.")
+	remoteHitTrackerTimeout      = flag.Duration("cache.remote_hit_tracker.rpc_timeout", 5*time.Second, "The timout to use for gRPC requests to the remote cache-hit-tracking service.")
+	remoteHitTrackerPollInterval = flag.Duration("cache.remote_hit_tracker.update_interval", 250*time.Millisecond, "The time interval to wait between sending remote cache-hit-tracking RPCs.")
+	maxPendingHitsPerGroup       = flag.Int("cache.remote_hit_tracker.max_pending_hits_per_group", 100_000, "The maximum number of pending cache-hit updates to store in memory for a given group.")
+	remoteHitTrackerWorkers      = flag.Int("cache.remote_hit_tracker.workers", 1, "The number of workers to use to send asynchronous remote cache-hit-tracking RPCs.")
+)
+
+func Register(env *real_environment.RealEnv) error {
+	conn, err := grpc_client.DialInternal(env, *remoteHitTrackerTarget)
+	if err != nil {
+		return err
+	}
+	env.SetHitTrackerFactory(newHitTrackerClient(env, conn))
+	return nil
+}
+
+func newHitTrackerClient(env *real_environment.RealEnv, conn grpc.ClientConnInterface) *HitTrackerFactory {
+	factory := HitTrackerFactory{
+		authenticator:          env.GetAuthenticator(),
+		ticker:                 env.GetClock().NewTicker(*remoteHitTrackerPollInterval).Chan(),
+		quit:                   make(chan struct{}, 1),
+		maxPendingHitsPerGroup: *maxPendingHitsPerGroup,
+		hitsByGroup:            map[groupID]*cacheHits{},
+		rpcTimeout:             *remoteHitTrackerTimeout,
+		client:                 hitpb.NewHitTrackerServiceClient(conn),
+	}
+	for i := 0; i < *remoteHitTrackerWorkers; i++ {
+		go factory.start()
```

Comment:
nit: rename to `runWorker` since it both starts and waits for completion, and since it's run once per worker

Reply:
Done